File: src/main/java/zielu/gittoolbox/revision/RevisionDataProvider.java
package zielu.gittoolbox.revision;

import com.intellij.openapi.vcs.history.VcsRevisionNumber;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.openapi.vfs.newvfs.impl.NullVirtualFile;
import java.time.ZonedDateTime;
import org.jetbrains.annotations.Nullable;

public interface RevisionDataProvider {
  RevisionDataProvider EMPTY = new RevisionDataProvider() {
    @Nullable
    @Override
    public ZonedDateTime getAuthorDateTime(int lineIndex) {
      return null;
    }

    @Nullable
    @Override
    public String getAuthor(int lineIndex) {
      return null;
    }

    @Nullable
    @Override
    public String getAuthorEmail(int lineIndex) {
      return null;
    }

    @Nullable
    @Override
    public String getSubject(int lineIndex) {
      return null;
    }

    @Override
    public VcsRevisionNumber getRevisionNumber(int lineIndex) {
      return VcsRevisionNumber.NULL;
    }

    @Override
    public VcsRevisionNumber getBaseRevision() {
      return VcsRevisionNumber.NULL;
    }

    @Override
    public VirtualFile getFile() {
      return NullVirtualFile.INSTANCE;
    }

    @Override
    public int getLineCount() {
      return 0;
    }

    @Override
    public String toString() {
      return "RevisionDataProvider[EMPTY]";
    }
  };

  @Nullable
  ZonedDateTime getAuthorDateTime(int lineIndex);

  @Nullable
  String getAuthor(int lineIndex);

  @Nullable
  String getAuthorEmail(int lineIndex);

  @Nullable
  String getSubject(int lineIndex);

  VcsRevisionNumber getRevisionNumber(int lineIndex);

  VcsRevisionNumber getBaseRevision();

  VirtualFile getFile();

  int getLineCount();
}

File: src/main/kotlin/zielu/gittoolbox/revision/RevisionDataProvider.kt
package zielu.gittoolbox.revision

import com.intellij.openapi.vcs.history.VcsRevisionNumber
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.openapi.vfs.newvfs.impl.NullVirtualFile
import java.time.ZonedDateTime

internal interface RevisionDataProvider {
  val baseRevision: VcsRevisionNumber
  val file: VirtualFile
  val lineCount: Int

  fun getRevisionNumber(lineIndex: Int): VcsRevisionNumber

  fun getAuthorDateTime(lineIndex: Int): ZonedDateTime?

  fun getAuthor(lineIndex: Int): String?

  fun getAuthorEmail(lineIndex: Int): String?

  fun getSubject(lineIndex: Int): String?

  companion object {
    val empty = object : RevisionDataProvider {
      override val baseRevision: VcsRevisionNumber
        get() = VcsRevisionNumber.NULL
      override val file: VirtualFile
        get() = NullVirtualFile.INSTANCE
      override val lineCount: Int
        get() = 0

      override fun getRevisionNumber(lineIndex: Int): VcsRevisionNumber = VcsRevisionNumber.NULL

      override fun getAuthorDateTime(lineIndex: Int): ZonedDateTime? = null

      override fun getAuthor(lineIndex: Int): String? = null

      override fun getAuthorEmail(lineIndex: Int): String? = null

      override fun getSubject(lineIndex: Int): String? = null

      override fun toString(): String = "RevisionDataProvider[EMPTY]"
    }
  }
}
